src/Microsoft.AspNetCore.Razor.Language/Legacy/MarkupParser.cs

```diff
@@ -0,0 +1,916 @@
+                builder.Add(OutputAsMarkupLiteral());
+
+                var markup = SyntaxFactory.MarkupBlock(builder.ToList());
+
+                return SyntaxFactory.RazorDocument(markup);
+            }
+        }
+
+        private void ParseMarkupNodes(
+            in SyntaxListBuilder<RazorSyntaxNode> builder,
+            ParseMode mode,
+            Func<SyntaxToken, bool> stopCondition = null)
+        {
+            stopCondition = stopCondition ?? (token => false);
+            while (!EndOfFile && !stopCondition(CurrentToken))
+            {
+                ParseMarkupNode(builder, mode);
+            }
+        }
+
+        private void ParseMarkupNode(in SyntaxListBuilder<RazorSyntaxNode> builder, ParseMode mode)
+        {
+            switch (GetParserState(mode))
+            {
+                case ParserState.MarkupText:
+                    ParseMarkupText(builder);
+                    break;
+                case ParserState.Tag:
+                    ParseMarkupElement(builder);
+                    break;
+                case ParserState.SpecialTag:
+                    ParseSpecialTag(builder);
+                    break;
+                case ParserState.XmlPI:
+                    ParseXmlPI(builder);
+                    break;
+                case ParserState.CData:
+                    ParseCData(builder);
+                    break;
+                case ParserState.MarkupComment:
+                    ParseMarkupComment(builder);
+                    break;
+                case ParserState.RazorComment:
+                    ParseRazorCommentWithLeadingAndTrailingWhitespace(builder);
+                    break;
+                case ParserState.DoubleTransition:
+                    ParseDoubleTransition(builder);
+                    break;
+                case ParserState.CodeTransition:
+                    ParseCodeTransition(builder);
+                    break;
+                case ParserState.Misc:
+                    ParseMisc(builder);
+                    break;
+                case ParserState.Unknown:
+                    AcceptAndMoveNext();
+                    break;
+                case ParserState.EOF:
+                    builder.Add(OutputAsMarkupLiteral());
+                    break;
+            }
+        }
+
+        private void ParseMarkupText(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            AcceptAndMoveNext();
+        }
+
+        private void ParseMarkupElement(in SyntaxListBuilder<RazorSyntaxNode> builder)
+        {
+            Assert(SyntaxKind.OpenAngle);
+
+            // Output already accepted tokens if any.
+            builder.Add(OutputAsMarkupLiteral());
+
+            if (!NextIs(SyntaxKind.ForwardSlash))
+            {
+                // Parsing a start tag
+                var startTag = ParseStartTag(out var expectsBody);
+                builder.Add(startTag);
+
+                var tagName = GetTagName(startTag);
```

Comment:
You should put a tag name convenience property that's set on creation or helper method (whatever you feel is better) on the MarkupTagBlockSyntax, at which point you wouldn't have to pass around this tag name any more. It'll also help at the TagHelper phase later on

Reply:
That already exists https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Language/Syntax/SyntaxNodeExtensions.cs#L245. But that operates on Red nodes. In this case, we are operating on Green nodes. I could make this an extension method but I was waiting to see how much it's used and then do it later.